File: Yuna/chap2/Chap2_2/app/src/main/java/org/techtown/chap2_2/MainActivity.java
package org.techtown.chap2_2;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    TextView textView;
    EditText editText;
    Button button_send;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textView = (TextView) findViewById(R.id.textView);
        editText = (EditText) findViewById(R.id.editTextTextPersonName);
        button_send = (Button) findViewById(R.id.button);


        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String txt = editText.getText().toString();
                textView.setText(txt.length()+" / 80바이트");
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }
    public void sendButtonClicked(View v){
        Toast toastView = Toast.makeText(this, editText.getText(), Toast.LENGTH_SHORT);
        toastView.show();
    }



}

File: RetrofitDemo/app/src/main/java/org/techtown/retrofitdemo/MainActivity.kt
package org.techtown.retrofitdemo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.liveData
import retrofit2.Response

class MainActivity : AppCompatActivity() {
  private lateinit var retService: AlbumService
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    retService = RetrofitInstance
      .getRetrofitInstance()
      .create(AlbumService::class.java)
    //getRequestWithQueryParameters()
    //getRequestWithPathParameters()
    uploadAlbum()


  }

  private fun getRequestWithQueryParameters() {
    val responseLiveData: LiveData<Response<Albums>> = liveData {
      val response = retService.getSortedAlbums(3)
      emit(response)
    }
    responseLiveData.observe(this, Observer {
      val albumsList = it.body()?.listIterator()
      if (albumsList != null) {
        while (albumsList.hasNext()) {
          val albumsItem = albumsList.next()
          val result = " " + "Album TItle : ${albumsItem.title}" + "\n" +
                  " " + "Album id : ${albumsItem.id}" + "\n" +
                  " " + "User id : ${albumsItem.userId}" + "\n\n\n"
          findViewById<TextView>(R.id.text_view).append(result)
        }
      }
    })
  }

  private fun getRequestWithPathParameters() {
    //path parameter example
    val pathResponse: LiveData<Response<AlbumsItem>> = liveData {
      val response = retService.getAlbum(3)
      emit(response)
    }
    pathResponse.observe(this, Observer {
      val title = it.body()?.title
      Toast.makeText(applicationContext, title, Toast.LENGTH_SHORT).show()
    })
  }

  private fun uploadAlbum() {
    val album = AlbumsItem(0, "My title", 3)
    val postResponse: LiveData<Response<AlbumsItem>> = liveData {
      val response = retService.uploadAlbum(album)
      emit(response)
    }
    postResponse.observe(this, Observer {
      val receivedAlbumsItem = it.body()
      val result = " " + "Album TItle : ${receivedAlbumsItem?.title}" + "\n" +
              " " + "Album id : ${receivedAlbumsItem?.id}" + "\n" +
              " " + "User id : ${receivedAlbumsItem?.userId}" + "\n\n\n"
      findViewById<TextView>(R.id.text_view).append(result)
    })
  }
}

